File: app/src/main/java/com/example/data/repository/RecommendationEngine.kt
package com.example.data.repository

import com.example.data.model.VideoEntity
import com.example.util.YouTubeUtils

data class AlgorithmSettings(
    val creatorWeight: Float = 0.7f,        // 0.0 to 1.0 (Importance of favorite creators)
    val discoveryRatio: Float = 0.2f,       // 0.0 to 1.0 (Fresh discoveries vs familiar topics)
    val shortsMode: String = "Carousel",     // "Carousel", "Hidden", "Separate"
    val minDurationMinutes: Int = 0,        // 0, 3, 5, 10 minutes
    val freshnessDecay: String = "Medium",   // "Slow", "Medium", "Fast"
    val autoDeleteDownloads: String = "Never", // "Never", "24h", "48h", "7d", "30d", "Watched"
    val blockedKeywords: List<String> = emptyList(), // Custom keywords/channels permanently excluded
    val boostedTopics: List<String> = emptyList()     // Custom topics/creators prioritized at the top
)

object RecommendationEngine {

    // Boredom Detection state (tracked 100% locally on device)
    private var consecutiveShortSkips = 0

    fun recordVideoWatchedDuration(durationSec: Int) {
        if (durationSec < 20) {
            consecutiveShortSkips++
        } else {
            consecutiveShortSkips = 0
        }
    }

    val isBoredomModeActive: Boolean
        get() = consecutiveShortSkips >= 3

    /**
     * Determines user-facing transparency reason for why a video is recommended.
     */
    fun getRecommendationReason(
        video: VideoEntity,
        favorites: List<VideoEntity>,
        watchHistory: List<VideoEntity>
    ): String {
        if (video.lastPositionSeconds > 0) return "🕒 Continue Watching"
        if (video.isFavorite) return "⭐ Saved Favorite"
        if (video.isWatchLater) return "🔖 Saved Watch Later"

        val isSubscribedProfileChannel = com.example.data.model.WillRyanProfileData.subscribedChannels.any {
            it.contains(video.channelName, ignoreCase = true) || video.channelName.contains(it, ignoreCase = true)
        }
        if (isSubscribedProfileChannel) return "💡 Subscribed Channel"

        val hasWatchedChannel = watchHistory.any { it.channelName.equals(video.channelName, ignoreCase = true) }
        if (hasWatchedChannel) return "🔥 Channel You Enjoy"

        if (isBoredomModeActive) return "✨ Fresh Topic Exploration"

        return "📈 Popular Recommendation"
    }

    /**
     * Calculates personal relevance score for a list of videos based on user settings and activity.
     */
    fun scoreAndRankVideos(
        videos: List<VideoEntity>,
        favorites: List<VideoEntity>,
        watchHistory: List<VideoEntity>,
        mutedChannels: List<com.example.data.model.MutedChannelEntity> = emptyList(),
        dislikedVideoIds: Set<String> = emptySet(),
        settings: AlgorithmSettings
    ): List<VideoEntity> {
        if (videos.isEmpty()) return emptyList()

        val mutedNames = mutedChannels.map { it.channelName.lowercase() }.toSet()
        val blockedLower = settings.blockedKeywords.map { it.trim().lowercase() }.filter { it.isNotEmpty() }
        val boostedLower = settings.boostedTopics.map { it.trim().lowercase() }.filter { it.isNotEmpty() }

        val unmutedVideos = videos.filter { video ->
            if (video.youtubeId in dislikedVideoIds) return@filter false
            val titleLower = video.title.lowercase()
            val chanLower = video.channelName.lowercase()
            chanLower !in mutedNames &&
            !YouTubeUtils.isForeignLanguageContent(video.title, video.channelName) &&
            blockedLower.none { blk -> titleLower.contains(blk) || chanLower.contains(blk) }
        }

        // 1. Identify top favorite channels
        val topChannels = (favorites.map { it.channelName } + watchHistory.map { it.channelName })
            .groupingBy { it }
            .eachCount()

        // 2. Identify top watched keywords in titles
        val topKeywords = watchHistory
            .flatMap { it.title.lowercase().split("\\s+".toRegex()) }
            .filter { it.length > 3 && it !in setOf("video", "with", "this", "that", "from", "2026", "youtube") }
            .groupingBy { it }
            .eachCount()

        val scoredList = unmutedVideos.map { video ->
            var score = 50.0f

            // A. Favorite & Watch Later Boost
            if (video.isFavorite) score += 40.0f
            if (video.isWatchLater) score += 25.0f

            // B. User Boosted Topics & Creators (Top Priority)
            val titleLower = video.title.lowercase()
            val chanLower = video.channelName.lowercase()
            val catLower = video.category.lowercase()
            if (boostedLower.any { bst -> titleLower.contains(bst) || chanLower.contains(bst) || catLower.contains(bst) }) {
                score += 90.0f
            }

            // C. Creator & Subscribed Profile Channel Boost
            val isSubscribedProfileChannel = com.example.data.model.WillRyanProfileData.subscribedChannels.any {
                it.contains(video.channelName, ignoreCase = true) || video.channelName.contains(it, ignoreCase = true)
            }
            if (isSubscribedProfileChannel) {
                score += 80.0f * settings.creatorWeight
            }

            val channelHits = topChannels[video.channelName] ?: 0
            if (channelHits > 0) {
                score += (channelHits * 15.0f * settings.creatorWeight).coerceAtMost(50.0f)
            }

            // C. Keyword & Subject Affinity
            val titleWords = video.title.lowercase().split("\\s+".toRegex())
            var keywordHits = 0
            for (word in titleWords) {
                if (topKeywords.containsKey(word)) {
                    keywordHits += topKeywords[word] ?: 0
                }
            }
            if (keywordHits > 0) {
                score += (keywordHits * 4.0f).coerceAtMost(30.0f)
            }

            // D. Discovery & Serendipity Boost (Wild Discovery Mode)
            if (channelHits == 0 && !isSubscribedProfileChannel) {
                score += settings.discoveryRatio * 75.0f
            }

            // E. Boredom Detection Engine: Inject Fresh Unvisited Topics
            if (isBoredomModeActive && channelHits == 0 && !isSubscribedProfileChannel) {
                score += 65.0f // Boost fresh unknown topics to break recommendation fatigue
            }

            // F. Freshness & Recency Engine (Massive boost for today's new uploads)
            val timeText = video.publishedTimeText.lowercase()
            val recencyBonus = when {
                timeText.contains("min") || timeText.contains("hour") || timeText.contains("just now") || timeText.contains("moment") -> 200.0f
                timeText.contains("today") -> 150.0f
                timeText.contains("1 day") || timeText.contains("yesterday") -> 80.0f
                timeText.contains("day") -> 50.0f
                timeText.contains("week") -> 25.0f
                else -> 5.0f
            }
            score += recencyBonus

            // G. Duration Preference Filter
            if (settings.minDurationMinutes > 0 && !YouTubeUtils.isShortVideo(video)) {
                val durSec = parseDurationSeconds(video.durationText)
                if (durSec > 0 && durSec < settings.minDurationMinutes * 60) {
                    score -= 40.0f // Penalize videos under minimum preferred duration
                }
            }

            // H. Watched Progress / Deprioritize Already Watched Videos on Feed
            if (video.lastPositionSeconds > 0) {
                score -= 60.0f // Deprioritize already watched videos to keep discovery feed fresh
            }

            Pair(video, score)
        }

        // 3. Sort descending by score, breaking ties with actual upload time
        val sortedList = scoredList
            .sortedWith(
                compareByDescending<Pair<VideoEntity, Float>> { it.second }
                    .thenBy { YouTubeUtils.parsePublishedTimeToSeconds(it.first.publishedTimeText) }
            )
            .map { it.first }

        // 4. Channel Diversity Cap: Enforce max 3 videos per channel in top 15, prioritizing newest
        val finalDiverseList = mutableListOf<VideoEntity>()
        val channelCounts = mutableMapOf<String, Int>()

        for (v in sortedList) {
            val count = channelCounts.getOrDefault(v.channelName, 0)
            val isFresh = v.publishedTimeText.contains("hour", ignoreCase = true) || v.publishedTimeText.contains("min", ignoreCase = true)
            // Allow up to 4 fresh videos for active creators, 2 for older videos
            val maxAllowed = if (isFresh) 4 else 2
            if (finalDiverseList.size < 15 && count >= maxAllowed) {
                continue
            }
            channelCounts[v.channelName] = count + 1
            finalDiverseList.add(v)
        }

        return finalDiverseList
    }

    private fun parseDurationSeconds(dur: String): Int {
        val parts = dur.split(":")
        return try {
            when (parts.size) {
                2 -> parts[0].toInt() * 60 + parts[1].toInt()
                3 -> parts[0].toInt() * 3600 + parts[1].toInt() * 60 + parts[2].toInt()
                else -> 0
            }
        } catch (e: Exception) { 0 }
    }
}

File: windows/VixzDesktop/Services/RecommendationEngine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using VixzDesktop.Models;

namespace VixzDesktop.Services
{
    public class AlgorithmSettings
    {
        public float CreatorWeight { get; set; } = 0.7f;
        public float DiscoveryRatio { get; set; } = 0.2f;
        public List<string> BlockedKeywords { get; set; } = new List<string>();
        public List<string> BoostedTopics { get; set; } = new List<string>();
    }

    public static class RecommendationEngine
    {
        private static readonly AlgorithmSettings DefaultSettings = new AlgorithmSettings();

        public static string GetRecommendationReason(
            VideoItem video,
            List<VideoItem> favorites,
            List<VideoItem> watchHistory,
            List<string> subscribedChannels)
        {
            if (StorageService.IsDisliked(video.Id) || StorageService.IsDeleted(video.Id)) return "⛔ Deprioritized / Removed";
            var savedPos = StorageService.GetPlaybackPosition(video.Id);
            if (savedPos > 3) return "🕒 Continue Watching";
            if (video.IsFavorite) return "👍 Liked Favorite";
            if (video.IsWatchLater) return "🔖 Saved Watch Later";

            var isSubscribed = subscribedChannels.Any(c =>
                c.Contains(video.ChannelTitle, StringComparison.OrdinalIgnoreCase) ||
                video.ChannelTitle.Contains(c, StringComparison.OrdinalIgnoreCase));
            if (isSubscribed) return "💡 Subscribed Channel";

            var hasWatched = watchHistory.Any(v => v.ChannelTitle.Equals(video.ChannelTitle, StringComparison.OrdinalIgnoreCase));
            if (hasWatched) return "🔥 Channel You Enjoy";

            return "📈 Popular Recommendation";
        }

        public static List<VideoItem> ScoreAndRankVideos(
            List<VideoItem> videos,
            List<VideoItem> favorites,
            List<VideoItem> watchHistory,
            List<string>? subscribedChannels = null,
            AlgorithmSettings? settings = null)
        {
            if (videos == null || videos.Count == 0) return new List<VideoItem>();

            var currentSettings = settings ?? DefaultSettings;
            var subChannels = subscribedChannels ?? WillRyanProfileData.SubscribedChannels;
            var dislikedIds = new HashSet<string>(StorageService.Settings.DislikedVideoIds ?? new List<string>(), StringComparer.OrdinalIgnoreCase);
            var deletedIds = new HashSet<string>(StorageService.Settings.DeletedVideoIds ?? new List<string>(), StringComparer.OrdinalIgnoreCase);
            var dislikedChannels = new HashSet<string>(StorageService.Settings.DislikedChannels ?? new List<string>(), StringComparer.OrdinalIgnoreCase);

            var blockedLower = currentSettings.BlockedKeywords
                .Select(k => k.Trim().ToLowerInvariant())
                .Where(k => !string.IsNullOrEmpty(k))
                .ToList();

            var boostedLower = currentSettings.BoostedTopics
                .Select(b => b.Trim().ToLowerInvariant())
                .Where(b => !string.IsNullOrEmpty(b))
                .ToList();

            // 0. Filter out permanently disliked, deleted, and blocked videos
            var filteredVideos = videos.Where(v =>
            {
                if (dislikedIds.Contains(v.Id) || deletedIds.Contains(v.Id)) return false;
                var titleLower = v.Title.ToLowerInvariant();
                var chanLower = v.ChannelTitle.ToLowerInvariant();
                return !blockedLower.Any(blk => titleLower.Contains(blk) || chanLower.Contains(blk));
            }).ToList();

            // 1. Identify top favorite channels
            var topChannels = favorites.Select(f => f.ChannelTitle)
                .Concat(watchHistory.Select(w => w.ChannelTitle))
                .GroupBy(c => c, StringComparer.OrdinalIgnoreCase)
                .ToDictionary(g => g.Key, g => g.Count(), StringComparer.OrdinalIgnoreCase);

            // 2. Identify top watched keywords in titles
            var stopWords = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
            {
                "video", "with", "this", "that", "from", "2026", "youtube", "official"
            };

            var topKeywords = watchHistory
                .SelectMany(w => Regex.Split(w.Title.ToLowerInvariant(), @"\s+"))
                .Where(w => w.Length > 3 && !stopWords.Contains(w))
                .GroupBy(w => w, StringComparer.OrdinalIgnoreCase)
                .ToDictionary(g => g.Key, g => g.Count(), StringComparer.OrdinalIgnoreCase);

            var scoredList = new List<(VideoItem Video, float Score)>();

            foreach (var video in filteredVideos)
            {
                float score = 50.0f;

                // A. Favorite & Watch Later Boost
                if (video.IsFavorite) score += 50.0f;
                if (video.IsWatchLater) score += 25.0f;

                // B. User Boosted Topics & Creators (+90 pts)
                var titleLower = video.Title.ToLowerInvariant();
                var chanLower = video.ChannelTitle.ToLowerInvariant();
                if (boostedLower.Any(bst => titleLower.Contains(bst) || chanLower.Contains(bst)))
                {
                    score += 90.0f;
                }

                // C. Disliked Creator / Negative Score Penalty (-80 pts)
                if (dislikedChannels.Contains(video.ChannelTitle) || dislikedChannels.Any(dc => dc.Contains(video.ChannelTitle, StringComparison.OrdinalIgnoreCase)))
                {
                    score -= 80.0f; // Heavy penalty for creators with disliked videos
                }

                // D. Subscribed Profile Channel Boost (+80 * CreatorWeight)
                var isSubscribedProfileChannel = subChannels.Any(c =>
                    c.Contains(video.ChannelTitle, StringComparison.OrdinalIgnoreCase) ||
                    video.ChannelTitle.Contains(c, StringComparison.OrdinalIgnoreCase));
                if (isSubscribedProfileChannel)
                {
                    score += 80.0f * currentSettings.CreatorWeight;
                }

                if (topChannels.TryGetValue(video.ChannelTitle, out int channelHits) && channelHits > 0)
                {
                    score += Math.Min(channelHits * 15.0f * currentSettings.CreatorWeight, 50.0f);
                }

                // E. Keyword & Subject Affinity
                var titleWords = Regex.Split(titleLower, @"\s+");
                int keywordHits = 0;
                foreach (var word in titleWords)
                {
                    if (topKeywords.TryGetValue(word, out int hits))
                    {
                        keywordHits += hits;
                    }
                }
                if (keywordHits > 0)
                {
                    score += Math.Min(keywordHits * 4.0f, 30.0f);
                }

                // F. Discovery Boost
                if (channelHits == 0 && !isSubscribedProfileChannel)
                {
                    score += currentSettings.DiscoveryRatio * 75.0f;
                }

                // G. Watched Progress / Deprioritize Completed Videos on Discovery Feed
                var savedPos = StorageService.GetPlaybackPosition(video.Id);
                if (savedPos > 3)
                {
                    score -= 60.0f;
                }

                video.AlgorithmScore = score;
                video.RecommendationReason = GetRecommendationReason(video, favorites, watchHistory, subChannels);

                scoredList.Add((video, score));
            }

            return scoredList
                .Where(s => s.Score > 0) // Exclude items with negative net score from top recommendations
                .OrderByDescending(s => s.Score)
                .Select(s => s.Video)
                .ToList();
        }
    }
}
